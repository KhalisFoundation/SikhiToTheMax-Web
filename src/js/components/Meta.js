import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { withRouter } from 'react-router-dom';
import DatePicker from 'react-date-picker';
import CalendarIcon from './Icons/CalendarIcon';
import Chevron from './Icons/Chevron';
import Hour24 from './Icons/Hour24';
import { getSourceId, getWriter, getRaag } from '@/util/api/shabad';

import { isFalsy, toAngURL, toNavURL, dateMath } from '../util';
import { TEXTS, PAGE_NAME, FIRST_HUKAMNAMA_DATE } from '@/constants';
import ForwardIcon from './Icons/ForwardIcon';

/**
 *
 *
 * @export
 * @class Meta
 * @augments {React.PureComponent<MetaProps>}
 */
class Meta extends React.PureComponent {
  static defaultProps = {
    nav: {},
  };

  /**
   * @typedef {object} MetaProps
   * @property {ShabadContentTypes} type
   * @property {object} info
   * @property {array} translationLanguages
   * @property {array} transliterationLanguages
   * @property {object} nav
   * @property {boolean} isUnicode
   * @property {boolean} isArrowsHidden
   *
   * @static
   * @memberof Meta
   */
  static propTypes = {
    history: PropTypes.object.isRequired,
    type: PropTypes.string.isRequired,
    info: PropTypes.object.isRequired,
    translationLanguages: PropTypes.array.isRequired,
    transliterationLanguages: PropTypes.array.isRequired,
    isUnicode: PropTypes.bool.isRequired,
    nav: PropTypes.shape({
      current: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
      previous: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
      next: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
    }),
    isArrowsHidden: PropTypes.bool
  };

  renderLeftArrow() {
    const {
      type,
      nav = {},
      isArrowsHidden,
    } = this.props;

    const link = toNavURL(this.props);
    const isShowPreviousArrow = isFalsy(nav.previous) === false;
    const isSync = type === 'sync';
    const isHukamnama = type === 'hukamnama';

    if (!isArrowsHidden) {
      if (isShowPreviousArrow) {
        return (
          <div className="shabad-nav left">
            <Link to={link + nav.previous}>
              {isHukamnama ? (
                <div className='hukamnama-nav-icon'>
                  <Hour24 direction='previous' />
                  <span>{dateMath.expand(nav.previous, false)}</span>
                </div>
              ) : (
                  <Chevron direction={Chevron.DIRECTIONS.LEFT} />
                )}
            </Link>
          </div>
        )
      }
      if (!isSync) {
        return (
          <div className="shabad-nav left disabled-nav">
            <a>
              {isHukamnama ? (
                <Hour24 direction='previous' />
              ) : (
                  <Chevron direction={Chevron.DIRECTIONS.LEFT} />
                )}
            </a>
          </div>
        )
      }
    }

    return '';
  }


  renderRightArrow() {
    const {
      type,
      nav = {},
      isArrowsHidden,
    } = this.props;

    const isShowNextArrow = isFalsy(nav.next) === false
    const isSync = type === 'sync';
    const isHukamnama = type === 'hukamnama';

    if (!isArrowsHidden) {

      if (isShowNextArrow) {
        return (
          <div className="shabad-nav right">
            <a role="button" aria-label="next" onClick={this.goToNextAng}>
              {isHukamnama ? (
                <div className='hukamnama-nav-icon'>
                  <Hour24 direction='next' />
                  <span>{dateMath.expand(nav.next, false)}</span>
                </div>
              ) : (
                  <Chevron direction={Chevron.DIRECTIONS.RIGHT} />
                )}
            </a>
          </div>
        )
      }

      if (!isSync) {
        return (
          <div className="shabad-nav right disabled-nav">
            <a>
              {isHukamnama ? (
                <Hour24 direction='next' />
              ) : (
                  <Chevron direction={Chevron.DIRECTIONS.RIGHT} />
                )}
            </a>
          </div>
        )
      }
    }

    return '';
  }

  render() {
    const {
      type,
      info,
      nav,
      isUnicode,
      translationLanguages,
      transliterationLanguages,
    } = this.props;
    const Item = ({ children, last = false }) =>
      children ? (
        <React.Fragment>
          {children}
          {last ? '' : ' - '}
        </React.Fragment>
      ) : null;

    const shouldShowEnglishInHeader =
      translationLanguages.includes('english') ||
      transliterationLanguages.includes('english');
    const contentType = isUnicode ? 'unicode' : 'gurmukhi'
    const isHukamnama = type === 'hukamnama';

    // const isShabad = type === 'shabad';
    return (
      <div id="metadata" className={`metadata-${type}`}>
        {!isHukamnama && this.renderLeftArrow()}

        <div className="meta">
          {isHukamnama && (
            <div className="meta-hukamnama">
              <div className="meta-hukamnama-left">
                <DatePicker
                  clearIcon={null}
                  onChange={this.goToParticularHukamnama}
                  value={new Date(nav.current)}
                  maxDate={new Date()}
                  minDate={new Date(FIRST_HUKAMNAMA_DATE)}
                  calendarIcon={<CalendarIcon width={20} />}
                />
                <span>Past Hukamnamas</span>
              </div>
              <h4>
                {TEXTS.HUKAMNAMA}, <span>{nav.current}</span>
              </h4>
              <div className="meta-hukamnama-right">
                <ForwardIcon />
                <Link to={`/shabad?id=${info.shabadId}`}>
                  {TEXTS.GO_TO_SHABAD}
                </Link>
              </div>
            </div>
          )}
          <h4 className="gurbani-font">
            <Item>
              {getRaag(info) &&
                getRaag(info)[contentType] &&
                getRaag(info)[contentType] !== 'null' &&
                getRaag(info)[contentType]}
            </Item>
            <Item>{getWriter(info) && getWriter(info)[contentType]}</Item>
            <Item>{info.source[contentType]}</Item>
            {info.source.pageNo !== null && (
              <Item last>
                <Link
                  to={toAngURL({
                    ang: info.source.pageNo,
                    source: getSourceId(info),
                  })}
                >
                  {getSourceId(info) == 'G' ? PAGE_NAME.ANG[contentType] : PAGE_NAME.PANNA[contentType]}{' '}
                  {info.source.pageNo}
                </Link>
              </Item>
            )}
          </h4>

          {shouldShowEnglishInHeader && (
            <h4>
              <Item>
                {getRaag(info) &&
                  getRaag(info)['english'] &&
                  getRaag(info)['english'] !== 'null' &&
                  getRaag(info)['english']}
              </Item>
              <Item>{getWriter(info) && getWriter(info)['english']}</Item>
              <Item>{info.source.english}</Item>
              {info.source.pageNo !== null && (
                <Item last>
                  <Link
                    to={toAngURL({
                      ang: info.source.pageNo,
                      source: getSourceId(info),
                    })}
                  >
                    {getSourceId(info) == 'G' ? 'Ang' : 'Pannaa'}{' '}
                    {info.source.pageNo}
                  </Link>
                </Item>
              )}
            </h4>
          )}
        </div>

        {!isHukamnama && this.renderRightArrow()}
      </div>
    );
  }

  /**
   * Handle SaveAng
   * @memberof Meta
   */

  goToParticularHukamnama = (date) => {
    const month = date.getMonth() + 1;
    const year = date.getFullYear();
    const day = date.getDate();

    const hukamnamaDate = `${year}/${month}/${day}`;
    const link = toNavURL(this.props)
    this.props.history.push(link + hukamnamaDate);
  }
  goToNextAng = () => {
    const link = toNavURL(this.props);
    this.props.history.push(link + this.props.nav.next);
  };
}

export default withRouter(Meta);
