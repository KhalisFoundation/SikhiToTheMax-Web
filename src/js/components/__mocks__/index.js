export * from './shabad';
export * from './store';
