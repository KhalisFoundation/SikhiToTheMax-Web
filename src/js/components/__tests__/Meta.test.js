
import React from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import {
  render,
  screen,
} from '@testing-library/react';
import { toHaveClass } from '@testing-library/jest-dom/matchers'
expect.extend({ toHaveClass });

import { shabad, mockStore } from '../__mocks__/';
import Meta from '../Meta';

//NOTE: MemoryRouter - https://testing-library.com/docs/example-react-router#reducing-boilerplate

describe('<Meta />', () => {
  const translationLanguages = ['english', 'spanish', 'punjabi'];
  const transliterationLanguages = ['english'];
  const store = mockStore({ darkMode: false, autoScrollMode: false });

  describe('<Meta type=\"ang\" />', () => {
    let renderedComponent = null;

    beforeEach(() => {
      renderedComponent = render(
        <Provider store={store}>
          <Meta
            type="ang"
            isUnicode={false}
            info={shabad.shabadInfo}
            translationLanguages={translationLanguages}
            transliterationLanguages={transliterationLanguages} />
        </Provider >,
        { wrapper: MemoryRouter }
      );
    });

    it('renders correctly', () => {
      const { container } = renderedComponent;
      expect(container).toMatchSnapshot();
    })

    // it('shows navigation arrows properly on fullscreen mode', async () => {
    //   const { container } = renderedComponent;
    //   const fullscreen = container.querySelector('.fullscreen');
    //   await fireEvent.click(fullscreen);

    //   // Basic limitation with css style as we can't test style changes on nav items
    //   const body = document.querySelector('body');
    //   expect(body).toHaveClass('fullscreen-view');
    // });
  });

  describe('<Meta type=\"shabad\" />', () => {
    let renderedComponent = null;
    beforeEach(() => {
      renderedComponent = render(
        <Meta
          type="shabad"
          isUnicode={false}
          info={shabad.shabadInfo}
          translationLanguages={translationLanguages}
          transliterationLanguages={transliterationLanguages} />,
        { wrapper: MemoryRouter }
      );
    })

    it('renders correctly', () => {
      const { container } = renderedComponent;
      expect(container).toMatchSnapshot();
    })

    it('shows chevron navigation buttons', () => {
      const { queryAllByTestId } = renderedComponent;
      expect(queryAllByTestId('chevron')).toHaveLength(2);
    });
  });

  describe('<Meta type=\"hukamnama\" />', () => {
    let renderedComponent = null;
    beforeEach(() => {
      renderedComponent = render(
        <Provider store={store}>
          <Meta
            type="hukamnama"
            isUnicode={false}
            info={shabad.shabadInfo}
            translationLanguages={translationLanguages}
            transliterationLanguages={transliterationLanguages} />
        </Provider >,
        { wrapper: MemoryRouter }
      );
    })
    it('renders correctly', () => {
      const { container } = renderedComponent;
      expect(container).toMatchSnapshot();
    })

    it('shows hours navigation buttons', () => {
      const { queryAllByTestId } = renderedComponent;
      expect(screen.queryAllByTestId('hour24')).toHaveLength(2);
    });
  });
});
