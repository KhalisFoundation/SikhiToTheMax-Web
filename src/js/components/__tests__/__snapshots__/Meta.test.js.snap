// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Meta /> <Meta type="ang" /> renders correctly 1`] = `
<div>
  <div
    class="metadata-ang"
    id="metadata"
  >
    <div
      class="shabad-nav left disabled-nav"
    >
      <a>
        <svg
          class="chevron "
          data-testid="chevron"
          height="1em"
          style="transform: rotate(0deg);"
          viewBox="0 0 320 512"
          width="1em"
        >
          <path
            d="M34.52 239.03L228.87 44.69c9.37-9.37 24.57-9.37 33.94 0l22.67 22.67c9.36 9.36 9.37 24.52.04 33.9L131.49 256l154.02 154.75c9.34 9.38 9.32 24.54-.04 33.9l-22.67 22.67c-9.37 9.37-24.57 9.37-33.94 0L34.52 272.97c-9.37-9.37-9.37-24.57 0-33.94z"
          />
        </svg>
      </a>
    </div>
    <div
      class="meta"
    >
      <h4
        class="gurbani-font"
      >
        rwgu swrMg
         - 
        sRI gurU gRMQ swihb jI
         - 
        <a
          href="/ang?ang=1211&source=G"
        >
          AMg
           
          1211
        </a>
        
      </h4>
      <h4>
        Raag Saarang
         - 
        Sri Guru Granth Sahib Ji
         - 
        <a
          href="/ang?ang=1211&source=G"
        >
          Ang
           
          1211
        </a>
        
      </h4>
    </div>
    <div
      class="shabad-nav right disabled-nav"
    >
      <a>
        <svg
          class="chevron "
          data-testid="chevron"
          height="1em"
          style="transform: rotate(180deg);"
          viewBox="0 0 320 512"
          width="1em"
        >
          <path
            d="M34.52 239.03L228.87 44.69c9.37-9.37 24.57-9.37 33.94 0l22.67 22.67c9.36 9.36 9.37 24.52.04 33.9L131.49 256l154.02 154.75c9.34 9.38 9.32 24.54-.04 33.9l-22.67 22.67c-9.37 9.37-24.57 9.37-33.94 0L34.52 272.97c-9.37-9.37-9.37-24.57 0-33.94z"
          />
        </svg>
      </a>
    </div>
  </div>
</div>
`;

exports[`<Meta /> <Meta type="hukamnama" /> renders correctly 1`] = `
<div>
  <div
    class="metadata-hukamnama"
    id="metadata"
  >
    <div
      class="meta"
    >
      <div
        class="meta-hukamnama"
      >
        <div
          class="meta-hukamnama-left"
        >
          <div>
            react calendar
          </div>
          <span>
            Past Hukamnamas
          </span>
        </div>
        <h4>
          Daily Hukamnama Sri Harmandir Sahib, Amritsar
          , 
          <span />
        </h4>
        <div
          class="meta-hukamnama-right"
        >
          <svg
            style="fill: rgb(1, 102, 155); width: 18px;"
            viewBox="0 0 491.807 491.807"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M485.802 221.184l-184.32-184.32a20.506 20.506 0 00-22.323-4.444c-7.659 3.154-12.636 10.65-12.636 18.924v82.002c-117.842 4.833-220.651 84.398-253.583 198.41C-1.314 381.195-.208 421.356.201 436.491l.082 3.973c0 9.216 6.164 17.306 15.032 19.743 1.823.492 3.645.737 5.448.737a20.51 20.51 0 0017.592-9.994c81.121-136.356 188.6-140.227 228.168-136.376v105.411c0 8.274 4.977 15.77 12.636 18.924 7.68 3.195 16.466 1.413 22.323-4.444l184.32-184.32c8.007-8.01 8.007-20.954 0-28.961z"
            />
          </svg>
          <a
            href="/shabad?id=4299"
          >
            Go to shabad
          </a>
        </div>
      </div>
      <h4
        class="gurbani-font"
      >
        rwgu swrMg
         - 
        sRI gurU gRMQ swihb jI
         - 
        <a
          href="/ang?ang=1211&source=G"
        >
          AMg
           
          1211
        </a>
        
      </h4>
      <h4>
        Raag Saarang
         - 
        Sri Guru Granth Sahib Ji
         - 
        <a
          href="/ang?ang=1211&source=G"
        >
          Ang
           
          1211
        </a>
        
      </h4>
    </div>
  </div>
</div>
`;

exports[`<Meta /> <Meta type="shabad" /> renders correctly 1`] = `
<div>
  <div
    class="metadata-shabad"
    id="metadata"
  >
    <div
      class="shabad-nav left disabled-nav"
    >
      <a>
        <svg
          class="chevron "
          data-testid="chevron"
          height="1em"
          style="transform: rotate(0deg);"
          viewBox="0 0 320 512"
          width="1em"
        >
          <path
            d="M34.52 239.03L228.87 44.69c9.37-9.37 24.57-9.37 33.94 0l22.67 22.67c9.36 9.36 9.37 24.52.04 33.9L131.49 256l154.02 154.75c9.34 9.38 9.32 24.54-.04 33.9l-22.67 22.67c-9.37 9.37-24.57 9.37-33.94 0L34.52 272.97c-9.37-9.37-9.37-24.57 0-33.94z"
          />
        </svg>
      </a>
    </div>
    <div
      class="meta"
    >
      <h4
        class="gurbani-font"
      >
        rwgu swrMg
         - 
        sRI gurU gRMQ swihb jI
         - 
        <a
          href="/ang?ang=1211&source=G"
        >
          AMg
           
          1211
        </a>
        
      </h4>
      <h4>
        Raag Saarang
         - 
        Sri Guru Granth Sahib Ji
         - 
        <a
          href="/ang?ang=1211&source=G"
        >
          Ang
           
          1211
        </a>
        
      </h4>
    </div>
    <div
      class="shabad-nav right disabled-nav"
    >
      <a>
        <svg
          class="chevron "
          data-testid="chevron"
          height="1em"
          style="transform: rotate(180deg);"
          viewBox="0 0 320 512"
          width="1em"
        >
          <path
            d="M34.52 239.03L228.87 44.69c9.37-9.37 24.57-9.37 33.94 0l22.67 22.67c9.36 9.36 9.37 24.52.04 33.9L131.49 256l154.02 154.75c9.34 9.38 9.32 24.54-.04 33.9l-22.67 22.67c-9.37 9.37-24.57 9.37-33.94 0L34.52 272.97c-9.37-9.37-9.37-24.57 0-33.94z"
          />
        </svg>
      </a>
    </div>
  </div>
</div>
`;
